### DISCUSSION TITLE
Handling not-implemented _get_server_version_info() in mssql.base

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#7396] [Usage Questions] [answered]
### Describe the bug

This is probably a corner case.

The `MSDialect` dialect seems to assume `system_version_info` is always a tuple.
But the initialize() in `DefaultDialect` can set it to None if `_get_server_version_info()` is not implemented

### To Reproduce

```python
from sqlalchemy.dialects import registry
from sqlalchemy.dialects.mssql.base import MSDialect

class CustomDialect(MSDialect):
    @classmethod
    def dbapi(cls):
        import pymssql

        return pymssql

registry.register('custom', 'myapp.custom_dialects', 'CustomDialect')
```

### Error

```
  File "venv/lib/python3.7/site-packages/sqlalchemy/engine/create.py", line 688, in first_connect
    dialect.initialize(c)
  File "venv/lib/python3.7/site-packages/sqlalchemy/dialects/mssql/base.py", line 2838, in initialize
    self._setup_version_attributes()
  File "venv/lib/python3.7/site-packages/sqlalchemy/dialects/mssql/base.py", line 2852, in _setup_version_attributes
    if self.server_version_info[0] not in list(range(8, 17)):
TypeError: 'NoneType' object is not subscriptable
```

### Versions

- OS: Ubuntu 20.04
- Python: 3.7
- SQLAlchemy: 1.4.27
- Database: SQL Server
- DBAPI: pymssql

### Additional context

I was trying to implement a custom dbapi for the MSSQL, and was getting errors about `'NoneType' object is not subscriptable` for server_version_info in the base `MSDialect` dialect.

I went through the code and found that in:
 - In the base `MSDialect` - the `system_version_info` is assumed to be a tuple in most cases
 - But in `DefaultDialect` - the value could be None if `_get_server_version_info()` was NotImplemented

Now, in `MSDialect_pymssql` and `MSDialect_pyodbc` - things are fine though, as it does implement `_get_server_version_info()`

On further checking, I see that the implementation of `_get_server_version_info` for mssql+pymssql and mssql+pyodbc are different. One calls `select @@version` and the other calls `SELECT CAST(SERVERPROPERTY('ProductVersion') AS VARCHAR)`
I assume , both SQL queries should be able to run on both drivers ?

Maybe there is some scope to move the version-info logic to the MSDialect base class

**Comment by CaselIT** [comment, marked as answer]
Hi,

> I was trying to implement a custom dbapi for the MSSQL, and was getting errors about 'NoneType' object is not subscriptable for server_version_info in the base MSDialect dialect.

the expectation is that this method is implemented, since it may be used by the dialect to compile in a different way some queries.
You can return a static tuple if you don't have a working connection yet.

**Comment by AbdealiLoKo** [comment]
I see, I notice that the pymssql dialect will return None in some cases too - probably can be looked into
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mssql/pymssql.py#L101

If there is an expectation that it's implemented, maybe raising a NotImplementedError seems more appropriate than setting it to None - which masks the error

I don't mind either way - just thought I'd report it as it was confusing and I wouldn't be able to understand what was happening unless I went through the sqlalchemy source code

**Reply by CaselIT** [reply]
> I see, I notice that the pymssql dialect will return None in some cases too - probably can be looked into
> https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mssql/pymssql.py#L101

Not sure if there's a reason why it's that way, it may simply be that it was just implemented that way and there was never a reason to change it.

Regarding your other question, in principle yes, using NotImplementedError is usually better for public things, but the MSDialect is not a public interface from sqlalcehmy point of view, it's just the default internal implementation of a dialect used by sqlalchemy, so it's expected that to interface to it you need to look at the code and their subclasses to see what's required.
